enterprise/reporting/error_index_reporting.go

```diff
@@ -0,0 +1,160 @@
+)
+
+const (
+	ei = "ei"
+)
+
+type payload struct {
+	MessageID        string    `json:"messageId"`
+	SourceID         string    `json:"sourceId"`
+	DestinationID    string    `json:"destinationId"`
+	TransformationID string    `json:"transformationId"`
+	TrackingPlanID   string    `json:"trackingPlanId"`
+	FailedStage      string    `json:"failedStage"`
+	EventType        string    `json:"eventType"`
+	EventName        string    `json:"eventName"`
+	ReceivedAt       time.Time `json:"receivedAt"`
+	FailedAt         time.Time `json:"failedAt"`
+}
+
+type ErrorIndexReporter struct {
+	ctx              context.Context
+	log              logger.Logger
+	configSubscriber *configSubscriber
+	errIndexDB       *jobsdb.Handle
+	now              func() time.Time
+
+	config struct {
+		dsLimit              misc.ValueLoader[int]
+		skipMaintenanceError bool
+		jobRetention         time.Duration
+	}
+}
+
+func NewErrorIndexReporter(
+	ctx context.Context,
+	conf *config.Config,
+	log logger.Logger,
+	configSubscriber *configSubscriber,
+) *ErrorIndexReporter {
+	eir := &ErrorIndexReporter{
+		ctx:              ctx,
+		log:              log,
+		configSubscriber: configSubscriber,
+		now:              time.Now,
+	}
+
+	eir.config.dsLimit = conf.GetReloadableIntVar(0, 1, "Reporting.errorIndexReporting.dsLimit")
+	eir.config.skipMaintenanceError = conf.GetBool("Reporting.errorIndexReporting.skipMaintenanceError", false)
+	eir.config.jobRetention = conf.GetDurationVar(24, time.Hour, "Reporting.errorIndexReporting.jobRetention")
+
+	eir.errIndexDB = jobsdb.NewForReadWrite(
+		ei,
+		jobsdb.WithDSLimit(eir.config.dsLimit),
+		jobsdb.WithConfig(conf),
+		jobsdb.WithSkipMaintenanceErr(eir.config.skipMaintenanceError),
+		jobsdb.WithJobMaxAge(
+			func() time.Duration {
+				return eir.config.jobRetention
+			},
+		),
+	)
+	if err := eir.errIndexDB.Start(); err != nil {
+		panic(fmt.Sprintf("failed to start error index db: %v", err))
+	}
+
+	return eir
+}
+
+// Report reports the metrics to the ei JobsDB
+func (eir *ErrorIndexReporter) Report(metrics []*types.PUReportedMetric, _ *sql.Tx) error {
+	jobs := lo.Flatten(lo.FilterMap(metrics, func(metric *types.PUReportedMetric, _ int) ([]*jobsdb.JobT, bool) {
+		if metric.StatusDetail == nil {
+			return nil, false
+		}
+
+		var jobs []*jobsdb.JobT
+		for _, failedMessage := range metric.StatusDetail.FailedMessages {
+			workspaceID := eir.configSubscriber.WorkspaceIDFromSource(metric.SourceID)
+
+			payload := payload{
+				MessageID:        failedMessage.MessageID,
+				SourceID:         metric.SourceID,
+				DestinationID:    metric.DestinationID,
+				TransformationID: metric.TransformationID,
+				TrackingPlanID:   metric.TrackingPlanID,
+				FailedStage:      metric.PUDetails.PU,
+				EventName:        metric.StatusDetail.EventName,
+				EventType:        metric.StatusDetail.EventType,
+				ReceivedAt:       failedMessage.ReceivedAt,
+				FailedAt:         eir.now(),
```

Comment:
**minor:** let's use the same time for all messages in the batch, i.e. calculate `failedAt := eir.now()` before starting the loop